### DISCUSSION TITLE
implement clangd's `textDocument/switchSourceHeader`

### DISCUSSION THREAD

**Discussion by markblunk** [joaotavora/eglot#1420] [Show and tell]
I wrote a function that uses `clangd`'s extension [textDocument/switchSourceHeader](https://clangd.llvm.org/extensions#switch-between-sourceheader) in eglot.

```elisp
(defun eglot--clangd-switch-source-header ()
  "Switch between the corresponding C/C++ source and header file.
Only works with clangd."
  (interactive)
  (let ((response
         (eglot--request
          (eglot--current-server-or-lose)
          :textDocument/switchSourceHeader (eglot--TextDocumentIdentifier))))
    (if response
;;        (funcall #'find-file (eglot-uri-to-path response))
        (find-file (eglot-uri-to-path response))
      (user-error "Could not find other file"))))
```

This is clearly influenced by similar functionality in lsp-mode https://github.com/emacs-lsp/lsp-mode/pull/2372/files

Since this is not in the official LSP spec, I am not sure if there is interest in including this in the eglot code base. Also I'm sure it needs some more robust error handling.

**Comment by nemethf** [comment]
I think the built-in `ff-find-other-file` for c files is superior to textDocument/switchSourceHeader. 

Additionally,  it would be slightly better not to define a new command (eglot--clangd-switch-source-header), but to extend the functionality of ff-find-other-file.  There are examples to extend ff-find-other-file in eglot-x for rust and toml language servers.

**Reply by joaotavora** [reply]
Non LSP-extension don't belong in Eglot anyway, perhaps @nemethf's own eglot-x.el though.  But the point about `ff-find-other-file` is a good one.

Anyway, I will try this as I mostly use clangd myself and it fits snugly in my `.emacs`.  BTW, why the `funcall`?

**Reply by markblunk** [reply]
1. `funcall` is there because of an ignorant copy and paste from the `lsp-mode` implementation. Thanks for pointing that out. Any other suggestions on how to improve the implementation? I don't have a lot of experience with elisp.

2. I did not know about `ff-find-other-file`. I will look into it. At a cursory glance it seems like it doesn't work out of the box, particularly when the header and source files are in different directories. With clangd you are able to make use of the `compile_commands.json` file to find the corresponding source/header file.

3. `eglot-x` sounds like a great place for this.

**Reply by joaotavora** [reply]
> With clangd you are able to make use of the compile_commands.json file to find the corresponding source/header file.

(this is completely besides the Eglot discussion, but akshually -- without knowing exactly how clangd does it -- I don't think it's just compile_commands.json.  That just teaches clangd how translation units are compiled individually and nothing of its includes, just the include paths. if I had to guess it's a probably a heuristic based on file names, amount of inline code, etc....  But it should definitely be better than `ff-find-other-file` unless I'm missing a lot of magic)

**Reply by nemethf** [reply]
>  it should definitely be better than `ff-find-other-file` 

I liked ff-find-other-file better because of its behavior when `ff-ignore-include` is nil, i.e., it can also open #included files.   It turns out it is possible to extend ff-find-other-file while keeping this behavior.  So I'm going to implement the support for textDocument/switchSourceHeader in eglot-x.
